### PR TITLE
Visualising waypoints3

### PR DESCRIPTION
**Pull request by forcecore**
Github doesn't seem to be picking up my new push and I'm not sure how to reopen #13336. With Parent-child implementation of Move activity in bleed, now we can work on waypoint again.

### CODE REVIEW

File: OpenRA.Mods.Cnc/Traits/MadTank.cs [hunk 1]
```diff
@@ -189,9 +189,11 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 				if (!order.Queued)
 					self.CancelActivity();
 
-				self.SetTargetLine(target, Color.Red);
+				self.ShowTargetLines();
 				self.QueueActivity(new MoveAdjacentTo(self, target));
-				self.QueueActivity(new CallFunc(StartDetonationSequence));
+				var det = new CallFunc(StartDetonationSequence);
+				det.SetTargetLineNode(target, Color.Red, true);
```
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/MadTank.cs]: Why isn't this already handled by the `MoveAdjacentTo`?

File: OpenRA.Game/Activities/Activity.cs [hunk 2]
```diff
@@ -265,6 +281,11 @@ public virtual IEnumerable<Target> GetTargets(Actor self)
 		{
 			yield break;
 		}
+
+		public virtual TargetLineNode? TargetLineNode(Actor self)
+		{
+			return null;
```
Comment (pchote) [inline OpenRA.Game/Activities/Activity.cs]: The comment above explains to create a `TargetLineNode` with `Target.Invalid` for non-drawables, which conflicts with returning null here?

File: OpenRA.Game/Activities/CallFunc.cs [hunk 3]
```diff
@@ -29,5 +33,15 @@ public override Activity Tick(Actor self)
 			if (a != null) a();
 			return NextActivity;
 		}
+
+		public void SetTargetLineNode(Target target, Color color, bool isTerminal)
```
Comment (pchote) [inline OpenRA.Game/Activities/CallFunc.cs]: Can we please default `bool isTerminal = false` so that most of the activities don't need to declare that?

File: OpenRA.Mods.Common/Activities/Air/HeliFly.cs [hunk 4]
```diff
@@ -61,9 +62,6 @@ public override Activity Tick(Actor self)
 				return NextActivity;
 			}
 
-			if (IsCanceled || !target.IsValidFor(self))
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/HeliFly.cs]: Why was this removed?

File: OpenRA.Game/Activities/Activity.cs (hunk 2 above)
Comment (pchote) [reply OpenRA.Game/Activities/Activity.cs]: But an even better option (which i've brought up on earlier PRs) is to make this `public virtual IEnumerable<TargetLineNode> TargetLineNodes(Actor self)` to match `GetTargets` and allow `yield break`ing on activities that don't want to return anything.

File: OpenRA.Mods.Common/Activities/DeliverUnit.cs [hunk 5]
```diff
@@ -214,5 +214,16 @@ public override bool Cancel(Actor self, bool keepQueue = false)
 
 			return base.Cancel(self);
 		}
+
+		public override TargetLineNode? TargetLineNode(Actor self)
+		{
+			if (innerActivity != null && innerActivity is HeliFly)
+			{
+				var a = innerActivity as HeliFly;
+				return new TargetLineNode(a.Target, Color.Yellow, true);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/DeliverUnit.cs]: Why do you need to create this explicitly?  This should be able to chain to the inner activity directly.

File: OpenRA.Mods.Common/Traits/Attack/AttackSuicides.cs [hunk 6]
```diff
@@ -74,11 +74,12 @@ public void ResolveOrder(Actor self, Order order)
...
 					self.CancelActivity();
 
-				self.SetTargetLine(target, Color.Red);
-
 				self.QueueActivity(move.MoveToTarget(self, target));
+				self.ShowTargetLines();
 
-				self.QueueActivity(new CallFunc(() => self.Kill(self)));
+				var act = new CallFunc(() => self.Kill(self));
+				act.SetTargetLineNode(target, Color.Red, true);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Attack/AttackSuicides.cs]: Likewise here: the `MoveToTarget` should already be drawing the target line, so why repeat it?

File: OpenRA.Mods.Common/Traits/Carryall.cs [hunk 7]
```diff
@@ -247,7 +247,7 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 					if (!order.Queued)
 						self.CancelActivity();
 
-					self.SetTargetLine(target, Color.Yellow);
+					self.ShowTargetLines();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Carryall.cs]: Shouldn't this (and the cases in the files below) be done after the QueueActivity?

File: OpenRA.Mods.Cnc/Traits/MadTank.cs (hunk 1 above)
Comment (forcecore) [reply OpenRA.Mods.Cnc/Traits/MadTank.cs]: I don't get this question? MoveAdjacentTo is supposed to be drawn in green (or not drawn at all) so I'm painting the detonation activity red. This activity appears when you order a Mad tank to queue a move order then queue the detonation order by right-clicking on the target structure.

File: OpenRA.Mods.Common/Activities/Air/HeliFly.cs (hunk 4 above)
Comment (forcecore) [reply OpenRA.Mods.Common/Activities/Air/HeliFly.cs]: I was confused while rebase-merging: target field became Target. I'll put this one back in.

File: OpenRA.Mods.Common/Activities/DeliverUnit.cs (hunk 5 above)
Comment (forcecore) [reply OpenRA.Mods.Common/Activities/DeliverUnit.cs]: I wasn't confident with chaining as I only saw this activity used in conjunction with HeliFly activity. I'll have it chained.

File: OpenRA.Mods.Common/Traits/Attack/AttackSuicides.cs (hunk 6 above)
Comment (forcecore) [reply OpenRA.Mods.Common/Traits/Attack/AttackSuicides.cs]: MoveToTarget is supposed to be used in conjunction with many activities so sometimes, it is red (C4 bombing), sometimes it is yellow (engi capture) so this is not repeating.

File: OpenRA.Mods.Common/Traits/Carryall.cs (hunk 7 above)
Comment (forcecore) [reply OpenRA.Mods.Common/Traits/Carryall.cs]: I just replaced the lines that go SetTargetLine to ShowTargetLines so the changes are easier to see. The target line drawing happens for some ticks so that it will display correct activity queue after one tick. On the other hand, you are right about this.

File: OpenRA.Mods.Cnc/Traits/MadTank.cs (hunk 1 above)
Comment (forcecore) [reply OpenRA.Mods.Cnc/Traits/MadTank.cs]: Now I get what you mean, seeing the comment below. As mentioned below, MoveAdjacentTo isn't really drawn as the color of the line depends on the activity that comes after MoveAdjacentTo. (e.g., Demolition, ExternalCapture, Dock, etc.)

File: OpenRA.Mods.Common/Activities/Air/HeliFly.cs (hunk 4 above)
Comment (forcecore) [reply OpenRA.Mods.Common/Activities/Air/HeliFly.cs]: It is back in.

**Comment by MustaphaTR** [comment]
Needs Rebase.

**Comment by forcecore** [comment]
Rebased again.

**Comment by Smittytron** [comment]
A polish issue I see is that this pr highlights visually the frustration that players have with units having to complete their movement to the next cell before changing directions. While showing this extra cell move through waypoints is true, it also just looks bad.

Assuming the movement issue isn't going to be looked at any time soon, is there a way to cut this step visually and only show the commands the player has given?

**Comment by xan2622** [comment]
Hi @forcecore . 

I have just tested your [visualising_waypoints3](https://github.com/forcecore/OpenRA/tree/) branch but I noticed that waypoints disappear after 2 seconds.   :confused:

I don't know if it is done this way on purpose but I think this delay is too short (5 seconds should be better) but IMO, I think that waypoints should never disappear after the unit is selected.  Is there a consensus about this delay? 

EDIT: It seems that pressing ALT shows the waypoints (thanks @pchote for this tip). I think that this hotkey (ALT) should be added on the waypoint tooltip.

**Comment by forcecore** [comment]
@xan2622 This PR mimics RA2 style waypoints where you press down Z key (which is like the shift key in ORA) the waypoint is shown and it disappears immediately after Z goes up.

File: OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs [hunk 8]
```diff
@@ -161,6 +161,11 @@ public override IEnumerable<Target> GetTargets(Actor self)
 			return Target.None;
 		}
 
+		public override IEnumerable<TargetLineNode> TargetLineNodes(Actor self)
+		{
+			yield break;
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs]: Can you please pass the desired colour in via the constructor so that you can render the line here?
You can then remove the ugly workarounds that are needed to fix the line rendering on the next activity in the queue.

**Comment by Smittytron** [comment]
> Anyway, how can I reproduce your issue?

Just have some units moving and order them to change direction (without holding shift). They'll have small lines that go to the next cell.

File: OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs (hunk 8 above)
Comment (forcecore) [reply OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs]: I doubt this is a walkaround once the activities are sorted out. MoveAdjacentTo is likely to be a child activity that isn't supposed to be traversed by DrawLinetoTarget.cs, I think.

https://github.com/OpenRA/OpenRA/pull/14384/files/c37d0abc59d875fc206edd092ce53f5c686c3bb0#diff-d7a57eebc0a2b3d470e099590d38922eR69

Then line 69 of DrawLineToTarget.cs can safely become for (; a != null; a = a.NextInQueue).

**Comment by Arular101** [comment]
Please don't forget to update the copyright notice year.

**Comment by forcecore** [comment]
Copyright year and Move-order-without-shift addressed.
I tried to detect if the canceled Move activity is in canceled state but it never becomes canceled, so I detected if the path is empty instead.

File: OpenRA.Mods.Common/Activities/GuardTargetActivity.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
```
Comment (Arular101) [inline OpenRA.Mods.Common/Activities/GuardTargetActivity.cs]: I think something went wrong with updating this.

**Comment by forcecore** [comment]
@Arular101 Git did it right, I missed out the newly added file.

File: OpenRA.Mods.Common/Activities/GuardTargetActivity.cs (hunk 9 above)
Comment (forcecore) [reply OpenRA.Mods.Common/Activities/GuardTargetActivity.cs]: Git did it right, I missed out the newly added file.

File: OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs (hunk 8 above)
Comment (pchote) [reply OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs]: Likely but not guaranteed.  Making `MoveAdjacentTo` differently to most other activities and then adding a workaround for that in `CallFunc` is almost certainly going to turn into a maintenance problem looking forward.

**Comment by TheChosenEvilOne** [comment]
Any news on this?

**Comment by reaperrr** [comment]
This has been sitting idle for ages and surely won't be picked up before some time after next release, so closing and moving back to Future milestone to clean up PR queue.

**Comment by tovl** [comment]
Is anyone still planning to pick this up before next release? Otherwise I would be willing to take over. With almost all orders being queueable now and the extensive activity rework this would be a nice capstone.

**Comment by pchote** [comment]
I was planning to, but even better if you want to take this on.
